posts/2021/09/hypothesis-array-api.ipynb

```diff
@@ -0,0 +1,940 @@
+    "!pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "43dd3731-a8d2-47b3-b88e-10eddc645f34",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "!pip install ipytest\n",
+    "import ipytest; ipytest.autoconfig(display_columns=80)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "4c7c2561-0671-4763-b5e0-935a1957b68b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "!pip install rich\n",
+    "from rich import print"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "2f5d822b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import warnings; warnings.filterwarnings(\"ignore\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "06545d86-904a-44ca-83c5-d5304a251d55",
+   "metadata": {},
+   "source": [
+    "## What the Array API enables\n",
+    "\n",
+    "The [API](https://data-apis.org/array-api/latest/) standardises functionality of array libraries, which has [numerous benefits](https://data-apis.org/array-api/latest/use_cases.html) for both developers and users.\n",
+    "If you've used NumPy before you'll feel right at home!\n",
+    "\n",
+    "The most exciting prospect for me\n",
+    "is being able to easily write an array-consuming method\n",
+    "that works all the adopting libraries.\n",
+    "Let's try writing this method\n",
+    "to calculate the cumulative sums of an array—"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "893afcdb-66cb-4d54-b9b0-071c3f411e82",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def cumulative_sums(x):\n",
+    "    \"\"\"Return the cumulative sums of the elements.\"\"\"\n",
+    "    xp = x.__array_namespace__()\n",
+    "    \n",
+    "    result = xp.empty(x.size, dtype=x.dtype)\n",
+    "    result[0] = x[0]\n",
+    "    for i in range(1, x.size):\n",
+    "        result[i] = result[i - 1] + x[i]\n",
+    "        \n",
+    "    return result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ed669b65",
+   "metadata": {},
+   "source": [
+    "The all-important\n",
+    "[`__array_namespace__()`](https://data-apis.org/array-api/latest/API_specification/array_object.html#method-array-namespace) method\n",
+    "allows array-consuming methods to get the array's respective Array API module.\n",
+    "Conventionally we assign it to the variable `xp`.\n",
+    "\n",
+    "From there you just need to rely on the guarantees of the Array API\n",
+    "to support NumPy, TensorFlow, PyTorch, CuPy, etc.\n",
+    "all in one simple method!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6e824f0f",
+   "metadata": {},
+   "source": [
+    "## Good ol' unit tests\n",
+    "\n",
+    "I hope you'd want write some tests at some point 😉\n",
+    "\n",
+    "We can import NumPy's Array API implementation\n",
```

Comment:
Perhaps for the reader that is not too familiar with the Array API and its implementations, it might be worth including a quick sentence on why we are not importing pure NumPy here. What makes `nxp.asarray` different from `np.asarray`?